api/apps/geoprocessing/src/modules/cleanup-tasks/cleanup-tasks.service.ts

```diff
@@ -0,0 +1,232 @@
+        scenarioIds,
+        CHUNK_SIZE_FOR_BATCH_DB_OPERATIONS,
+      ).entries()) {
+        await entityManager.insert(
+          'scenario_nuke_preparation',
+          summaryChunks.map((chunk: string) => ({
+            scenario_id: chunk,
+          })),
+        );
+      }
+
+      const missingScenarioIdsFromScenarioFeaturesData: entitiesWithScenarioId[] = await this.geoEntityManager.query(
+        `SELECT sfd.scenario_id
+          FROM scenario_features_data sfd
+          LEFT JOIN scenario_nuke_preparation snp ON snp.scenario_id = sfd.scenario_id
+          WHERE snp.scenario_id IS NULL`,
+      );
+      const missingScenarioIdsFromBlmFinalResults: entitiesWithScenarioId[] = await this.geoEntityManager.query(
+        `SELECT bfr.scenario_id
+          FROM blm_final_results bfr
+          LEFT JOIN scenario_nuke_preparation snp ON snp.scenario_id = bfr.scenario_id
+          WHERE snp.scenario_id IS NULL`,
+      );
+      const missingScenarioIdsFromBlmPartialResults: entitiesWithScenarioId[] = await this.geoEntityManager.query(
+        `SELECT bpr.scenario_id
+          FROM blm_final_results bpr
+          LEFT JOIN scenario_nuke_preparation snp ON snp.scenario_id = bpr.scenario_id
+          WHERE snp.scenario_id IS NULL`,
+      );
+      const missingScenarioIdsFromMarxanExecutionMetadata: entitiesWithScenarioId[] = await this.geoEntityManager.query(
+        `SELECT mem.scenarioId
+          FROM marxan_execution_metadata mem
+          LEFT JOIN scenario_nuke_preparation snp ON snp.scenario_id = mem.scenarioId
+          WHERE snp.scenario_id IS NULL`,
+      );
+      const missingScenarioIdsFromScenariosPuData: entitiesWithScenarioId[] = await this.geoEntityManager.query(
+        `SELECT spd.scenario_id
+          FROM scenarios_pu_data spd
+          LEFT JOIN scenario_nuke_preparation snp ON snp.scenario_id = spd.scenario_id
+          WHERE snp.scenario_id IS NULL`,
+      );
+
+      const notMatchingScenarioIds: string[] = [];
+
+      // As opposed to Project cleanup, scenarios are flattened to their ids
+      // so the map just uses them to remove unused resources
+      notMatchingScenarioIds.push(
+        ...missingScenarioIdsFromScenarioFeaturesData.map((p) => p.scenario_id),
+        ...missingScenarioIdsFromBlmFinalResults.map((p) => p.scenario_id),
+        ...missingScenarioIdsFromBlmPartialResults.map((p) => p.scenario_id),
+        ...missingScenarioIdsFromMarxanExecutionMetadata.map(
+          (p) => p.scenario_id,
+        ),
+        ...missingScenarioIdsFromScenariosPuData.map((p) => p.scenario_id),
+      );
+
+      await notMatchingScenarioIds.map(async (scenarioId) => {
+        await this.scenarioUnusedResources.removeUnusedResources(scenarioId);
+        return scenarioId;
+      });
+
+      await this.geoEntityManager.query(
+        `TRUNCATE TABLE scenario_nuke_preparation`,
+      );
+    });
+  }
+
+  @Cron(CronExpression.EVERY_1ST_DAY_OF_MONTH_AT_MIDNIGHT)
```

Comment:
it's very much ok to hardcode the interval here, though I would still advise to move this to a `config` item, so that

- it's easier to find settings that admins may want to tweak in the future
- it can be overridden per-environment, in case admins so wish (prod vs staging vs testing, etc - the testing one may be useful in fact)

In order to do so we'd likely need to use the textual cron syntax, but I'd say that's ok.

And by all means, I would default to at the very most once a day, not more than that - but ideally every six hours or so.

Not because I expect this cleanup task to find so many dangling objects so frequently - these should all be an exception - but because in case of restarts of the Node process(es) due to new versions being deployed, restarts after failures, etc. especially close to the moment the scheduled cron event is going to become due in the event loop, we may risk skipping it and waiting until the next scheduled occurrence - so a monthly occurrence may be problematic.

On the other hand, almost all the tables we scan for dangling items have an index on `project_id` or `scenario_id` (as relevant), except `blm_<partial|final>_results`, but I don't think this index is going to be used in a negative constraint (`not in...`), so it _may_ still make sense to use the set difference query you used in your implementation, _but then_ use the resulting set to `delete ... where <scenario|project>_id in (<the set of scenario_ids not belonging to any current scenario>)` - but this would need some digging into query planning at scale... I'd not be too concerned at this stage 😏